### PR TITLE
[countersyncd]: resolve labels via object ids

### PR DESCRIPTION
**Pull request by Pterosaur**
## What I did
- pass `object_ids` through `SwssActor -> IPFixTemplatesMessage -> IpfixActor`
- store `object_ids` alongside `object_names` per session key
- resolve IPFIX labels via `label -> object_id -> object_name`
- keep the legacy positional fallback only when `object_ids` are unavailable
- update unit/integration tests for the new message shape

## Why
`object_ids` and `object_names` are positional pairs, while the IPFIX label matches the object ID rather than the 1-based index into `object_names`. The old logic incorrectly assumed:

`label -> object_names[label - 1]`

The correct logic is:

`label -> find matching object_id -> use the object_name at the same position`

## Validation
Ran in CI-like `sonic-slave-bookworm` container with Azure Pipelines deb artifacts installed:
- `cargo test --locked --manifest-path crates/countersyncd/Cargo.toml test_sai_stat_from_ipfix_uses_object_ids_mapping -- --nocapture`
- `cargo test --locked --manifest-path crates/countersyncd/Cargo.toml test_session_update_with_object_names -- --nocapture`
- `cargo test --locked --manifest-path crates/countersyncd/Cargo.toml test_session_update_without_object_names -- --nocapture`
- `cargo test --locked --manifest-path crates/countersyncd/Cargo.toml test_object_names_follow_template_id -- --nocapture`

### CODE REVIEW

File: crates/countersyncd/src/message/saistats.rs [hunk 1]
```diff
@@ -89,12 +91,13 @@ impl SAIStat {
...
+        // Resolve object name from label via object_ids.
+        let object_name = if let Some(index) = object_ids.iter().position(|object_id| *object_id == label) {
+            object_names
+                .get(index)
+                .cloned()
+                .unwrap_or_else(|| format!("unknown_{}", label))
         } else {
-            // Fallback to label number if object name not found
             format!("unknown_{}", label)
         };
```
[the remark this answers is not in the document]
Comment (Pterosaur) [reply crates/countersyncd/src/message/saistats.rs]: This is intentional. We explicitly removed the positional fallback because `label -> object_name` by index was the incorrect behavior that this PR is fixing. When `object_ids` are unavailable, the implementation now returns `unknown_<label>` instead of guessing from `object_names` order.

File: crates/countersyncd/src/message/saistats.rs [hunk 2]
```diff
@@ -89,12 +91,13 @@ impl SAIStat {
...
-        // Resolve object name from label
-        let object_name = if label > 0 && (label as usize) <= object_names.len() {
-            // Convert 1-based label to 0-based index
-            object_names[(label - 1) as usize].clone()
+        // Resolve object name from label via object_ids.
+        let object_name = if let Some(index) = object_ids.iter().position(|object_id| *object_id == label) {
+            object_names
+                .get(index)
+                .cloned()
+                .unwrap_or_else(|| format!("unknown_{}", label))
```
[the remark this answers is not in the document]
Comment (Pterosaur) [reply crates/countersyncd/src/message/saistats.rs]: Fixed in `fc66bf48`. `IpfixActor` now precomputes a per-template `object_id -> object_name` lookup map when template metadata is updated, and `SAIStat::from_ipfix()` does an O(1) lookup instead of scanning `object_ids` with `position()` for every field.

File: crates/countersyncd/src/actor/ipfix.rs [hunk 3]
```diff
@@ -501,6 +501,10 @@ pub struct IpfixActor {
     applied_templates_map: HashMap<String, Vec<u16>>,
     /// Mapping from message key to object names for converting label IDs
     object_names_map: HashMap<String, Vec<String>>,
+    /// Mapping from message key to object IDs aligned positionally with object_names
+    object_ids_map: HashMap<String, Vec<u16>>,
```
[the remark this answers is not in the document]
Comment (Pterosaur) [reply crates/countersyncd/src/actor/ipfix.rs]: Fixed in 1e64c0d8. `IpfixActor` now keeps `object_id_name_map` as the single source of truth for label-to-name resolution and removes the unused `object_names_map` / `object_ids_map` state to avoid redundant cloning and bookkeeping.

File: crates/countersyncd/src/actor/ipfix.rs [hunk 4]
```diff
@@ -621,10 +627,31 @@ impl IpfixActor {
...
+
+        if let (Some(object_names), Some(object_ids)) = (&templates.object_names, &templates.object_ids) {
+            self.object_id_name_map.insert(
+                templates.key.clone(),
+                object_ids
+                    .iter()
+                    .copied()
+                    .zip(object_names.iter().cloned())
+                    .collect(),
+            );
```
[the remark this answers is not in the document]
Comment (Pterosaur) [reply crates/countersyncd/src/actor/ipfix.rs]: Fixed in 1e64c0d8. The map build now validates `object_ids.len() == object_names.len()` before inserting, logs a warning on mismatch, and drops the stale lookup instead of relying on `zip()` truncation.

**Review by banidoru — APPROVED** [review]
Clean change that replaces fragile positional (1-based index) object name resolution with an explicit object_id-to-name HashMap lookup. Key observations:

**Correctness:** The precomputed `object_id_name_map` in `IpfixActor` correctly builds a `HashMap<u16, String>` from paired `object_ids`/`object_names`, with proper length validation and warning on mismatch. The `SAIStat::from_ipfix` lookup is now O(1) and gracefully falls back to `unknown_<label>`.

**Robustness:** `swss.rs` parsing handles invalid `object_ids` tokens well — breaks on first parse failure, warns, and clears the entire vec rather than silently dropping entries. Double validation (swss.rs + ipfix.rs) provides defense in depth.

**Design:** Using `std::collections::HashMap` for the inner map (vs `ahash::HashMap` used elsewhere) is a minor inconsistency but reasonable given the small map size and `u16` keys.

**Tests:** Good coverage including the new `test_sai_stat_from_ipfix_uses_object_ids_mapping` test with non-sequential IDs (10, 20). All existing tests properly updated.

LGTM.

File: crates/countersyncd/src/actor/ipfix.rs [hunk 5]
```diff
@@ -621,10 +621,27 @@ impl IpfixActor {
...
+                    object_ids
+                        .iter()
+                        .copied()
+                        .zip(object_names.iter().cloned())
+                        .collect(),
+                );
+            } else {
+                warn!(
+                    "IPFIX template object_ids/object_names length mismatch for key {}: ids={}, names={}. Skipping object_id_name_map entry.",
+                    templates.key,
```
Comment (banidoru) [inline crates/countersyncd/src/actor/ipfix.rs:651]: If `object_ids` contains duplicate entries (e.g. `[1, 1]` with names `["Eth0", "Eth1"]`), the `HashMap::collect()` will silently keep only the last mapping (`1 → "Eth1"`), dropping the first. Depending on whether IPFIX producers can ever emit duplicate object IDs per session, this could silently lose a counter's object name.

Consider either:
- Documenting that duplicate object IDs are undefined behavior, or
- Adding a warning log when a duplicate is detected during map construction (check `insert` return value).

Low severity — likely not a real-world scenario, but worth a defensive note.

**Review by banidoru — COMMENTED** [review]
Well-structured fix. The migration from positional label indexing to explicit object-ID-based lookup is correct, and previous review feedback (O(1) precomputation, zip truncation guard, unused map removal) has been addressed. One minor concern below about duplicate object IDs.

File: crates/countersyncd/src/actor/ipfix.rs [hunk 6]
```diff
@@ -499,8 +499,8 @@ pub struct IpfixActor {
     temporary_templates_map: HashMap<u16, String>,
     /// Mapping from message key to template IDs for applied templates
     applied_templates_map: HashMap<String, Vec<u16>>,
-    /// Mapping from message key to object names for converting label IDs
-    object_names_map: HashMap<String, Vec<String>>,
+    /// Precomputed lookup from object ID/label to object name for O(1) stat resolution
```
Comment (banidoru) [inline crates/countersyncd/src/actor/ipfix.rs:502]: Nit: `object_id_name_map` is declared as `StdHashMap` (std) while `temporary_templates_map` and `applied_templates_map` use `ahash::HashMap`. Since this map is consulted on the per-field hot path during data record processing, using `ahash::HashMap` here too would be consistent and marginally faster for small key counts. Non-blocking.

File: crates/countersyncd/src/message/saistats.rs [hunk 7]
```diff
@@ -325,16 +329,29 @@ mod tests {
...
+        let object_name_lookup = build_object_name_lookup(&[(1, "Ethernet0")]);
 
-        let stat = SAIStat::from_ipfix(&field_spec, &value, &object_names);
+        let stat = SAIStat::from_ipfix(&field_spec, &value, Some(&object_name_lookup));
 
         assert_eq!(stat.object_name, "unknown_0"); // Fallback for zero label
         assert_eq!(stat.type_id, 1);
         assert_eq!(stat.stat_id, 2);
         assert_eq!(stat.counter, 1000);
     }
```
Comment (banidoru) [inline crates/countersyncd/src/message/saistats.rs:342]: Minor test coverage gap: there's no test that calls `SAIStat::from_ipfix` with `object_name_lookup: None`. All existing tests pass `Some(...)`. A small test confirming the `None` path returns `unknown_<label>` would cover this branch explicitly. Non-blocking.

**Review by banidoru — APPROVED** [review]
Clean change. The shift from positional 1-based indexing to explicit object_id→name mapping via a precomputed HashMap is correct and well-structured. Previous review feedback (parse validation, O(1) lookup, unused map removal, zip-length guard) has all been addressed.

Two minor non-blocking observations:

1. `object_id_name_map` uses `std::collections::HashMap` while adjacent maps use `ahash::HashMap`. Both work, but the inconsistency is worth noting — if ahash was chosen for performance on the hot path, this map (which *is* on the hot path) would also benefit from it.

2. No unit test exercises `SAIStat::from_ipfix` with `object_name_lookup: None` directly. The existing tests all pass `Some(...)`. A one-liner test confirming `None` produces `unknown_<label>` would strengthen coverage of that branch.

**Review by banidoru — APPROVED** [review]
All reviewers approved. LGTM.

File: crates/countersyncd/src/actor/ipfix.rs (hunk 5 above)
Comment (banidoru) [reply crates/countersyncd/src/actor/ipfix.rs:651]: Addressed in f3a879da. The map construction now checks `insert()` return value for duplicates, logs a warning with both names, and drops the entire lookup to avoid ambiguous resolution. Good defensive handling.

File: crates/countersyncd/src/actor/ipfix.rs (hunk 6 above)
Comment (banidoru) [reply crates/countersyncd/src/actor/ipfix.rs:502]: On re-review: `object_id_name_map` is declared as `HashMap<String, HashMap<u16, String>>` where `HashMap` is `ahash::HashMap` (imported at line 1). Both the outer and inner maps already use ahash. This concern does not apply to the current code.

File: crates/countersyncd/src/message/saistats.rs (hunk 7 above)
Comment (banidoru) [reply crates/countersyncd/src/message/saistats.rs:342]: Addressed. New test `test_sai_stat_from_ipfix_without_lookup_uses_unknown_label` explicitly covers the `None` path and asserts `unknown_20`.

**Review by banidoru — COMMENTED** [review]
Re-review (iteration 2): All prior concerns have been thoroughly addressed.

**Resolved in new commits:**
- `f3a879da`: Duplicate object ID detection with `insert()` return-value check, warning log, and map removal — good defensive handling.
- `1e64c0d8`: Removed unused `object_names_map`/`object_ids_map`, length mismatch validation with warning.
- `fc66bf48`: Precomputed `HashMap<u16, String>` for O(1) label→name resolution.
- `39b8c676`: Strict `object_ids` parsing in `SwssActor` — invalid tokens abort the entire parse (no silent drops), length mismatch against `object_names` caught early.

**New commits reviewed — no new issues found.** The validation is now layered (SwssActor parse-time + IpfixActor map-build-time), object ID→name resolution is O(1), and edge cases (duplicates, mismatches, missing data) all degrade gracefully with logging.

Test coverage is solid: new tests for the `None` lookup path, object ID mapping, and existing tests updated for the new message shape.

LGTM.

**Review by banidoru — APPROVED** [review]
Re-review (iteration 2): All prior concerns addressed — duplicate object_id detection with map invalidation, length mismatch validation, ahash consistency confirmed, and None-path test coverage added. New diff is clean: parsing is strict (fail-fast on invalid tokens), validation is layered (SwssActor + IpfixActor), O(1) lookup is correct, and defensive cleanup on missing metadata is solid. No new issues found. LGTM.

**Review by banidoru — APPROVED** [review]
Re-review (iteration 2): All prior concerns have been addressed.

**Duplicate object_id handling** (ipfix.rs:651): Now checks `insert()` return value, logs warning, and drops the lookup on duplicates. Good defensive handling.

**ahash consistency** (ipfix.rs:502): Confirmed `object_id_name_map` already uses `ahash::HashMap` via the module import. Original concern was incorrect — no change needed.

**Test coverage for None lookup path** (saistats.rs:342): New test `test_sai_stat_from_ipfix_without_lookup_uses_unknown_label` covers the `None` branch explicitly.

**New diff review**: The object_ids parsing in swss.rs properly validates each token, fails fast on parse errors with a warning, and cross-checks length against object_names. The precomputed HashMap in IpfixActor is well-structured with proper cleanup on deletion and mismatch. No new issues found.

LGTM.

**Review by banidoru — APPROVED** [review]
All reviewers approved. LGTM.

**Review by banidoru — APPROVED** [review]
Re-review (iteration 3): All prior review concerns have been addressed. Key improvements since last review:

- Duplicate object ID detection with map invalidation (f3a879da)
- Test coverage for `None` lookup path added
- Clean validation chain: parsing validates tokens individually, length mismatch checked at both SwssActor and IpfixActor levels

The object_id→object_name resolution logic is correct, well-defended against edge cases (duplicates, length mismatches, parse failures), and properly tested. No new issues found. LGTM.

**Review by banidoru — APPROVED** [review]
All prior concerns addressed. The implementation is clean and well-validated:

- O(1) label→name lookup via precomputed HashMap — good.
- Duplicate object ID detection with map removal — defensive and correct.
- Length mismatch validation at both SwssActor and IpfixActor layers (defense-in-depth).
- Invalid token parsing in SwssActor aborts gracefully with warning.
- Test coverage added for the None-lookup path and object_ids mapping.

No new issues found. LGTM.

**Review by banidoru — APPROVED** [review]
Re-review (iteration 3): All prior concerns have been addressed. The implementation is solid — multi-level validation (SwssActor parse-time + IpfixActor map-build), duplicate object_id detection with safe fallback, length mismatch guards, and proper stale-state cleanup. New test coverage for the None lookup path is good. No new issues found. LGTM.

**Review by banidoru — APPROVED** [review]
All reviewers approved. LGTM.

**Comment by Pterosaur** [comment]
Created 202412 backport PR: https://github.com/Azure/sonic-swss.msft/pull/213

**Review by banidoru — APPROVED** [review]
All prior review threads resolved. Latest commit is a merge of master with no changes to PR files. LGTM.

**Review by banidoru — APPROVED** [review]
- All 9 prior review threads addressed and resolved
- Length mismatch validation, duplicate object ID detection, and unused state cleanup all look correct
- O(1) lookup via precomputed HashMap is clean; test coverage includes None path
- No new issues found in this iteration
- LGTM

**Review by banidoru — APPROVED** [review]
All reviewers approved. LGTM.

**Review by banidoru — APPROVED** [review]
Re-review (iteration 5) — all prior threads resolved, new diff LGTM.

- object_ids parsing in swss.rs now validates each token strictly, logs+aborts on invalid entries — good
- Length mismatch guard and duplicate object_id detection with map drop are solid defensive measures
- Unused object_names_map/object_ids_map removed; single object_id_name_map is clean
- Test coverage added for None lookup path and non-sequential object ID mapping
- All call sites updated consistently (benches, integration tests)

**Review by banidoru — APPROVED** [review]
- All prior review concerns addressed: O(1) precomputed lookup, length mismatch validation, duplicate ID detection, unused state removal
- `object_ids` parsing in `swss.rs` properly validates tokens and fails safely on invalid input
- Clean separation: validation at parse boundary (swss), map construction at template processing (ipfix), O(1) resolution at stat creation (saistats)
- Good test coverage including `None` lookup path and non-sequential object ID mapping
- LGTM

**Review by banidoru — APPROVED** [review]
- All prior review threads addressed and resolved — length mismatch validation, duplicate object_id detection, unused map removal, O(1) lookup precomputation all landed cleanly
- `swss.rs` object_ids parsing handles invalid tokens gracefully with early-exit and warning, plus cross-validates against object_names length before propagating
- `IpfixActor` correctly drops stale lookups on template deletion, mismatch, and duplicate scenarios — no orphaned state
- Test coverage is comprehensive: new tests for object_id mapping, None lookup path, and updated call sites across benches and integration tests
- Clean, well-structured iteration — LGTM

**Review by banidoru — APPROVED** [review]
All reviewers approved. LGTM.

**Review by banidoru — APPROVED** [review]
- All prior review concerns addressed: duplicate object_id detection, length mismatch validation, invalid token handling with warn+break, O(1) lookup via precomputed HashMap, and test coverage for `None` lookup path.
- Defense-in-depth validation: length mismatch checked in both `SwssActor` (parse time) and `IpfixActor` (apply time) — good practice.
- Clean removal of old positional (1-based index) resolution in favor of explicit object_id→name mapping.
- `object_id_name_map` properly cleaned up on all error/mismatch paths (remove from map, not stale data).
- Test coverage is thorough: new tests for object_id mapping, `None` lookup fallback, and integration tests updated.

**Review by banidoru — APPROVED** [review]
- All prior review concerns addressed and resolved
- Object ID → name lookup is correctly precomputed as O(1) HashMap
- Duplicate object_id and length-mismatch validation looks solid with proper fallback behavior
- Double validation in swss.rs (parse-time) and ipfix.rs (map-build) is belt-and-suspenders but acceptable
- Test coverage is comprehensive: covers None lookup, object_id mapping, and existing edge cases

**Review by banidoru — APPROVED** [review]
- All prior review threads resolved — duplicate object_id handling, length mismatch validation, O(1) lookup, test coverage for `None` path all addressed.
- Object ID parsing in `swss.rs` properly aborts on invalid tokens instead of silently dropping — good.
- Duplicate object_id detection in `ipfix.rs` correctly falls back to removing the map entry — prevents ambiguous resolution.
- Clean separation: validation at ingestion (SwssActor), precomputation at template processing (IpfixActor), O(1) lookup at stat resolution (SAIStat).
- LGTM — no new issues found.

**Review by banidoru — APPROVED** [review]
All reviewers approved. LGTM.

**Review by prsunny — APPROVED** [review]
